README.md
# LeProxy

LeProxy is the HTTP/SOCKS proxy server for everybody!

LeProxy is designed for anonymous surfing, improved security and privacy plus
circumventing geoblocking.
It allows you to enjoy the web like it's meant to work and access your favorite
online video platform without annoying country blocks while traveling.

LeProxy is a powerful, lightweight, fast and simple to use proxy server that you
can host on your own server or PC at home and then access from anywhere.
It supports optional authentication so you can share a server instance with your
family and friends without having to worry about third parties.
It provides compatibility with a large number of clients and services by
accepting both common HTTP and SOCKS proxy protocols on a single listening port.

**Table of contents**

* [Install](#install)
* [Usage](#usage)
* [Clients](#clients)
* [Development](#development)
* [License](#license)

> Note that this is an early beta version and that
  LeProxy is under active development.
  Many new features are going to be added in the future!

## Install

LeProxy requires only PHP.
*PHP 7+ is highly recommended*, but it runs on any system that uses PHP 5.4+ or
HHVM.
If you have not installed PHP already, on a recent Ubuntu/Debian system, simply run:

```bash
$ sudo apt-get install php7.0-cli
```

You can simply download the latest `leproxy-{version}.php` file from our
[releases page](https://github.com/leproxy/leproxy/releases).
The [latest release](https://github.com/leproxy/leproxy/releases/latest) can
always be downloaded like this:

```bash
$ curl -OL https://leproxy.org/leproxy-latest.php
```

Downloaded the `leproxy-{version}.php` file?
You did it!! Really simple, huh?

> LeProxy is distributed as a single PHP file that contains everything you need
  to run LeProxy.
  The below examples assume you have saved this file as `leproxy.php` locally,
  but you can use any name you want.
  If you're interested in the more technical details of this file, you may want
  to check out the [development instructions](#development) below.

## Usage

Once [installed](#install), you can start LeProxy by simply running:

```bash
$ php leproxy.php 
```

By default, LeProxy will listen on the public address `0.0.0.0:8080`. 
If you want to listen on another address, you can pass an explicit listening
address.
LeProxy will report an error if it fails to listen on the given address,
you may try another address or use port `0` to pick a random free port.
For example, if you do not want to allow accessing LeProxy from the outside and
only want to listen on the local interface:

```bash
$ php leproxy.php 127.0.0.1:8080
```

> The listening address MUST be in the form `ip:port` or just `ip` or `:port`,
  with the above defaults being applied (or you can use a Unix domain socket
  path).

Note that LeProxy runs in protected mode by default, so that it only forwards
requests from the local host and can not be abused as an open proxy.
If you have ensured only legit users can access your system, you can
pass the `--allow-unprotected` flag to forward requests from all hosts.
If you want to require the client to send username/password authentication
details, you can include this as part of the listening address:

```bash
$ php leproxy.php username:password@0.0.0.0:8080
```

> If the username or password contains special characters, make sure to use
  URL encoded values (percent-encoding) such as `p%40ss` for `p@ss`.

By default, LeProxy allows connections to every destination address as given in
each incoming proxy request.
If you want to block access to certain destination hosts and/or ports, you may
blacklist these by passing the `--block=<destination>` argument.
Any number of destination addresses can be given.
Each destination address can be in the form `host:port` or just `host` or
`:port` and `host` may contain the `*` wildcard to match anything.
Subdomains for each host will automatically be blocked.
For example, the following can be used to block access to `youtube.com` (and its
subdomains such as `www.youtube.com`) and port 80 on all hosts (standard
plaintext HTTP port):

```bash
$ php leproxy.php --block=youtube.com --block=*:80
```

> Note that the block list operates on the destination addresses as given in the
  incoming proxy request. Some [clients](#clients) use local DNS resolution and
  do not transmit hostnames, but only the resolved destination IP addresses
  (particularly common for the SOCKS protocol).
  Make sure to configure your client to use remote DNS resolution accordingly
  and/or also block access to relevant IP addresses.

As an alternative to listing each blocked destination as an individual command
line argument, you may also pass a path to a `hosts` file instead.
You can either create a hosts file mapping yourself if you only want to block
certain hosts or you may one of the many great hosts files.
For example, you can download a hosts file from https://github.com/StevenBlack/hosts
("hosts files from several well-curated sources like adaway.org, mvps.org,
malwaredomainlist.com, someonewhocares.org, and potentially others")
to use this as a very simple, yet effective adblocker.
Note that LeProxy will only block domains (and all its subdomains) that match
the IP `0.0.0.0` and ignore all other entries:

```bash
$ cat hosts.txt
127.0.0.1 localhost
0.0.0.0 ads.example.com
0.0.0.0 socialnetwork.example
…

$ php leproxy.php --block-hosts=hosts.txt
```

By default, Leproxy creates a direct connection to the destination address for
each incoming proxy request.
In this mode, the destination doesn't see the original client address, but only
the address of your LeProxy instance.
If you want a higher level degree of anonymity, you can use *proxy forwarding*,
where the connection will be tunneled through another upstream proxy server.
This may also be useful if your upstream proxy address changes regularly (such
as when using public proxies), but you do not want to reconfigure your client
every time or if your upstream proxy requires a feature that your client does
not support, such as requiring authentication or a different proxy protocol.
You can simply pass your upstream proxy server address as another URL parameter
after the listening address like this:

```bash
$ php leproxy.php --proxy=socks://user:pass@127.0.0.1:8080
```

> The upstream proxy server URI MUST contain a hostname or IP and SHOULD include
  a port unless the proxy happens to use default port `8080` (or you can use a
  Unix domain socket path).
  If no scheme is given, the `http://` scheme will be assumed.
  If no port is given, port `8080` will be assumed regardless of scheme.
  The `http://` and `socks[5]://` schemes support optional username/password
  authentication as in the above example.

By appending additional upstream proxy servers, this can effectively be turned
into *proxy chaining*, where each incoming proxy request will be forwarded
through the chain of all upstream proxy servers from left to right.
This comes at the price of increased latency, but may provide a higher level
degree of anonymity, as each proxy server in the chain only sees its direct
communication partners and the destination only sees the last proxy server in
the chain:

```bash
$ php leproxy.php --proxy=127.1.1.1:8080 --proxy=127.2.2.2:8080 --proxy=127.3.3.3:8080
```

By default, LeProxy prints a log message for every connection attempt to the
console output (STDOUT) for debugging and analysis purposes.
For privacy reasons, it does not persist (store) these log messages on its own.
If you do not want LeProxy to log anything, you may also pass the `--no-log`
flag.
If you want to persist the log to a log file, you may use standard operating
system facilities such as `tee` to redirect the output to a file:

```bash
$ php leproxy.php | tee -a leproxy.log
```

## Clients

Once LeProxy is running, you can start using it with pretty much any client
software.
The below example assumes you want to use a web browser, but LeProxy can
actually be used with any client software that provides proxy support, such as
an email or IM client.

Most clients provide settings to manually configure a proxy server in their
settings/preferences dialogs.
You can simply set the details from the listening address as configured above:

* Protocol: HTTP or SOCKS
* Server: 127.0.0.1 (or the public hostname or IP where LeProxy runs)
* Port: 8080

> Note that these settings have to be adjusted to your actual network settings.
  If you fail to provide correct settings, no further connection will succeed.
  In this case, simply remove or disable these settings again.
  The same may apply if you're roaming in another network or the proxy server is
  temporarily not available.

Many clients (in particular web browsers and mobile phones) also support Proxy
Auto-Configuration (PAC) by specifying a PAC URL.
Using PAC is often beneficial because most clients will simply ignore the proxy
settings if the PAC URL can not be reached, such as when you're roaming in
another network or the proxy server is temporarily not available.
Simply use the URL to your LeProxy instance in the following format:

```
http://127.0.0.1:8080/pac
```

> Note that these settings have to be adjusted to your actual network settings.
  If you fail to provide correct settings, you may or may not be able to
  establish further connections, as most clients will simply ignore invalid
  settings.
  If your client disallows this, simply remove or disable these settings again.
  LeProxy's PAC file instructs your client to use LeProxy as an HTTP proxy for
  all public HTTP requests.
  This means that hostnames that resolve to IPs from your local network will
  still use a direct connection without going through a proxy.

## Development

LeProxy is an [open-source project](#license) and encourages everybody to
participate in its development.
You're interested in checking out how LeProxy works under the hood and/or want
to contribute to the development of LeProxy?
Then this section is for you!

The recommended way to install LeProxy is to clone (or download) this repository
and use [Composer](http://getcomposer.org) to download its dependencies.
Therefore you'll need PHP, git and curl installed.
For example, on a recent Ubuntu/debian system, simply run:

```bash
$ sudo apt-get install php7.0-cli git curl
$ git clone https://github.com/leproxy/leproxy.git
$ cd leproxy
$ curl -s https://getcomposer.org/installer | php
$ sudo mv composer.phar /usr/local/bin/composer
$ composer install
```

That's it already!
You should now be able to run the development version of LeProxy simply by
running the `leproxy.php` file like this:

```bash
$ php leproxy.php
```

See also [usage](#usage) for more details.

LeProxy uses a sophisticated test suite for functional tests and integration
tests.
To run the test suite, go to the project root and run:

```bash
$ php vendor/bin/phpunit
```

If you want to distribute LeProxy as a single standalone release file, you may
compile the project into a single file like this:

```bash
$ php compile.php
```

> Note that compiling will temporarily uninstall all development dependencies
  for distribution and then re-install the complete set of dependencies.
  This should only take a second or two if you've previously installed its
  dependencies already.
  The compile script optionally accepts the version number (`VERSION` env) and
  an output file name or will otherwise try to look up the last release tag,
  such as `leproxy-1.0.0.php`.

In addition to the above test suite, LeProxy uses a simple bash/curl-based
acceptance test setup which can also be used to check the resulting release
file:

```bash
$ ./tests/acceptance.sh
```

> Note that the acceptance tests will try to locate a `leproxy*.php` file in
  the project directory to run the tests against. You may optionally supply the
  output file name to test against.

Made some changes to your local development version?

Make sure to let the world know! :shipit:
We welcome PRs and would love to hear from you!

Happy hacking!

## License

LeProxy is an open source project released under the permissive
[MIT license](LICENSE).

LeProxy is standing on the shoulders of giants.
Building something like LeProxy probably wouldn't be possible if not for the
excellent open source projects that it builds on top of.
In particular, it uses [ReactPHP](http://reactphp.org/) for its fast,
event-driven architecture.

All of its dependencies are managed through Composer, see also the
[development section](#development) for more details.
If you're using the [development version](#development), you may run
`$ composer licenses --no-dev` to get a list of all runtime dependencies and their
respective licenses.
All these requirements are bundled into the single standalone release file.


CHANGELOG.md
# Changelog

## 0.2.2 (2018-07-17)

*   Feature: Support listening on Unix domain socket (UDS) paths and
    support proxy forwarding/chaining via Unix domain socket (UDS) paths.
    (#51 by @clue)

    LeProxy now supports listening on Unix domain socket (UDS) paths and
    proxy forwarding/chaining via Unix domain socket (UDS) paths,
    both of which are considered advanced usage:

    ```bash
    $ php leproxy.php ./proxy.socket

    $ php.leproxy.php :8080 --proxy http+unix://./proxy.socket
    ```

*   Feature: Update HTTP dependencies and reject chunked requests and
    update ReactPHP to stable LTS releases.
    (#49 and #50 by @clue)

## 0.2.1 (2018-03-09)

*   Feature: Update Socket and DNS dependency to support loading system default
    DNS config on all supported platforms.
    (`/etc/resolv.conf` on Unix/Linux/Mac/Docker/WSL and WMIC on Windows)
    (#45 by @clue)

    This means that connecting to hosts that are managed by a local DNS server,
    such as a corporate DNS server or when using Docker containers, will now
    work as expected across all platforms with no changes required.

*   Fix: Update HTTP and HttpClient dependencies to include a number of
    improvements for HTTP handling (support multiple response cookies, larger
    request headers and ignore corrupt response Transfer-Encoding).
    (#46 by @clue)

*   Reduce package size by updating HttpClient dependency and removing unneeded deps.
    (#47 by @clue)

*   Improve test suite by adding forward compatibility with updated
    react/promise-stream and fix Travis builds by skipping all IPv6 tests.
    (#42 by @WyriHaximus and #44 by @clue)

## 0.2.0 (2017-09-01)

*   Feature: Add `--block=<target>` argument to blacklist destination addresses and
    add `--block-hosts=<path> argument` to block multiple hosts and
    use proper HTTP/SOCKS status codes and improve error reporting and analysis
    (#24, #40 and #41 by @clue)

    For example, the following command allows you to block all plaintext HTTP
    requests and use LeProxy as a simple, yet effective adblocker:

    ```bash
    $ php leproxy.php --block=:80 --block-hosts=hosts-ads.txt
    ```

*   Feature: Validate all arguments through commander instead of throwing exception
    (#37 by @clue)

*   Feature: Update Socket dependency to support hosts file on all platforms and
    update DNS dependency to fix Windows DNS timeout issues
    (#38 and #39 by @clue)

## 0.1.0 (2017-08-01)

* First tagged release :shipit:


leproxy.php
#!/usr/bin/env php
<?php
/**
 * LeProxy is the HTTP/SOCKS proxy server for everybody!
 *
 * LeProxy should be run from the command line. Assuming this file is
 * named `leproxy.php`, try running `$ php leproxy.php --help`.
 *
 * @link https://leproxy.org/ LeProxy project homepage
 * @license https://leproxy.org/#license MIT license
 * @copyright 2017 Christian Lück
 * @version dev
 */

namespace LeProxy\LeProxy;

use Clue\Commander\Router;
use Clue\Commander\NoRouteFoundException;
use Clue\Commander\Tokens\Tokenizer;
use React\EventLoop\Factory;
use React\Dns\Config\HostsFile;

if (PHP_VERSION_ID < 50308 || PHP_SAPI !== 'cli') {
    echo 'LeProxy HTTP/SOCKS proxy requires running ' . (PHP_SAPI !== 'cli' ? ('via command line (not ' . PHP_SAPI . ')') : ('on PHP 5.3.8+ (is ' . PHP_VERSION . ')')) . PHP_EOL;
    exit(1);
}

// get current version from git or default to "unknown" otherwise
// this line will be replaced with the static const in the release file.
define('VERSION', ltrim(exec('git describe --always --dirty 2>/dev/null || echo unknown'), 'v'));

require __DIR__ . '/vendor/autoload.php';

// parse options from command line arguments (argv)
$tokenizer = new Tokenizer();
$tokenizer->addFilter('block', function (&$value) {
    $value = ConnectorFactory::coerceBlockUri($value);
    return true;
});
$tokenizer->addFilter('proxy', function (&$value) {
    $value = ConnectorFactory::coerceProxyUri($value);
    return true;
});
$tokenizer->addFilter('hosts', function (&$value) {
    $value = HostsFile::loadFromPathBlocking($value)->getHostsForIp('0.0.0.0');
    return true;
});
$commander = new Router($tokenizer);
$commander->add('--version', function () {
    exit('LeProxy development version ' . VERSION . PHP_EOL);
});
$commander->add('-h | --help', function () {
    exit('LeProxy HTTP/SOCKS proxy

Usage:
    $ php leproxy.php [<listenAddress>] [--allow-unprotected] [--block=<destination>...] [--block-hosts=<path>...] [--proxy=<upstreamProxy>...] [--no-log]
    $ php leproxy.php --version
    $ php leproxy.php --help

Arguments:
    <listenAddress>
        The socket address to listen on.
        The address consists of a full URI which may contain a username and
        password, host and port.
        By default, LeProxy will listen on the public address 0.0.0.0:8080.
        LeProxy will report an error if it fails to listen on the given address,
        you may try another address or use port `0` to pick a random free port.
        If this address does not contain a username and password, LeProxy will
        run in protected mode and only forward requests from the local host,
        see also `--allow-unprotected`.

    --allow-unprotected
        If no username and password has been given, then LeProxy runs in
        protected mode by default, so that it only forwards requests from the
        local host and can not be abused as an open proxy.
        If you have ensured only legit users can access your system, you can
        pass the `--allow-unprotected` flag to forward requests from all hosts.
        This option should be used with care, you have been warned.

    --block=<destination>
        Blocks forwarding connections to the given destination address.
        Any number of destination addresses can be given.
        Each destination address can be in the form `host` or `host:port` and
        `host` may contain the `*` wildcard to match anything.
        Subdomains for each host will automatically be blocked.

    --block-hosts=<path>
        Loads the hosts file from the given file path and blocks all of the
        hostnames (and subdomains) that match the IP `0.0.0.0`.
        Any number of hosts files can be given, all hosts will be blocked.

    --proxy=<upstreamProxy>
        An upstream proxy server where each connection request will be
        forwarded to (proxy chaining).
        Any number of upstream proxies can be given.
        Each address consists of full URI which may contain a scheme, username
        and password, host and port. Default scheme is `http://`, default port
        is `8080` for all schemes.

    --no-log
        By default, LeProxy logs all connection attempts to STDOUT for
        debugging purposes. This can be avoided by passing this argument.

    --version
        Prints the current version of LeProxy and exits.

    --help, -h
        Shows this help and exits.

Examples:
    $ php leproxy.php
        Runs LeProxy on public default address 0.0.0.0:8080 (protected mode)

    $ php leproy.php 127.0.0.1:1080
        Runs LeProxy on custom address 127.0.0.1:1080 (protected mode, local only)

    $ php leproxy.php user:pass@0.0.0.0:8080
        Runs LeProxy on public default addresses and require authentication

    $ php leproxy.php --block=youtube.com --block=*:80
        Runs LeProxy on default address and blocks access to youtube.com and
        port 80 on all hosts (standard plaintext HTTP port).

    $ php leproxy.php --proxy=http://user:pass@127.1.1.1:8080
        Runs LeProxy so that all connection requests will be forwarded through
        an upstream proxy server that requires authentication.
');
});
$commander->add('[--allow-unprotected] [--block=<block:block>...] [--block-hosts=<file:hosts>...] [--proxy=<proxy:proxy>...] [--no-log] [<listen>]', function ($args) {
    // validate listening URI or assume default URI
    $args['listen'] = ConnectorFactory::coerceListenUri(isset($args['listen']) ? $args['listen'] : '');

    $args['allow-unprotected'] = isset($args['allow-unprotected']);
    if ($args['allow-unprotected'] && strpos($args['listen'], '@') !== false) {
        throw new \InvalidArgumentException('Unprotected mode can not be used with authentication required');
    }

    if (isset($args['block-hosts'])) {
        if (!isset($args['block'])) {
            $args['block'] = array();
        }
        foreach ($args['block-hosts'] as $hosts) {
            $args['block'] += $hosts;
        }
    }

    // filter duplicate block entries and subdomains
    if (isset($args['block'])) {
        $args['block'] = ConnectorFactory::filterRootDomains($args['block']);
    }

    return $args;
});
try {
    $args = $commander->handleArgv();
} catch (\Exception $e) {
    $message = '';
    if (!$e instanceof NoRouteFoundException) {
        $message = ' (' . $e->getMessage() . ')';
    }

    fwrite(STDERR, 'Usage Error: Invalid command arguments given, see --help' . $message . PHP_EOL);

    // sysexits.h: #define EX_USAGE 64 /* command line usage error */
    exit(64);
}

$loop = Factory::create();

// set next proxy server chain -> p1 -> p2 -> p3 -> destination
$connector = ConnectorFactory::createConnectorChain(isset($args['proxy']) ? $args['proxy'] : array(), $loop);

if (isset($args['block'])) {
    $connector = ConnectorFactory::createBlockingConnector($args['block'], $connector);
}

// log all connection attempts to STDOUT (unless `--no-log` has been given)
if (!isset($args['no-log'])) {
    $connector = new LoggingConnector($connector, new Logger());
}

// create proxy server and start listening on given address
$proxy = new LeProxyServer($loop, $connector);
try {
    $socket = $proxy->listen($args['listen'], $args['allow-unprotected']);
} catch (\RuntimeException $e) {
    fwrite(STDERR, 'Program error: Unable to start listening, maybe try another port? (' . $e->getMessage() . ')'. PHP_EOL);

    // sysexits.h: #define EX_OSERR 71 /* system error (e.g., can't fork) */
    exit(71);
}

$addr = str_replace('tcp://', 'http://', $socket->getAddress());
echo 'LeProxy HTTP/SOCKS proxy now listening on ' . $addr . ' (';
if (strpos($args['listen'], '@') !== false) {
    echo 'authentication required';
} elseif ($args['allow-unprotected']) {
    echo 'unprotected mode, open proxy';
} else {
    echo 'protected mode, local access only';
}
echo ')' . PHP_EOL;

if (isset($args['proxy'])) {
    echo 'Forwarding via: ' . implode(' -> ', $args['proxy']) . PHP_EOL;
}

if (isset($args['block'])) {
    echo 'Blocking a total of ' . count($args['block']) . ' destination(s)' . PHP_EOL;
}

$loop->run();
